### DISCUSSION TITLE
GPU install for JAX

### DISCUSSION THREAD

**Discussion by tgautam03** [jax-ml/jax#12337] [Q&A] [answered]
Hi

I've been trying to install jax with cuda support but can't figure out where I'm making a mistake. I manually installed nvidia drivers and CUDA (I just used CUDA toolkit and checked the nvidia drivers option while installing). I've driver version 515.65.01, CUDA version 11.7, and CUDNN version 8.5 (I got this from `nvidia-smi`).

I tried installing JAX using the command `pip install "jax[cuda11_cudnn82]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html` and I'm getting the following warning when I import jax.

```
2022-09-12 17:49:13.694508: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-09-12 17:49:13.719081: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-09-12 17:49:13.720791: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
```

I've no idea what's missing. It would be great if someone can help me out.

**Comment by jjyyxx** [comment]
I installed CUDA and CUDNN through NVidia's apt source. For me running `ldconfig -p | grep libcudart.so.11.0` prints
```
        libcudart.so.11.0 (libc6,x86-64) => /usr/local/cuda/targets/x86_64-linux/lib/libcudart.so.11.0
```
This is because I have an entry `000_cuda.conf` in `/etc/ld.so.conf.d/` which contains a single line `/usr/local/cuda/targets/x86_64-linux/lib`, so ld could find the libraries.

Your manual installation may cause this trouble. You could try to put the same file in `/etc/ld.so.conf.d/`, then rebuild cache with `ldconfig` and things may work as expected.

**Reply by tgautam03** [reply]
Did you use `sudo apt install nvidia-cuda-toolkit` to install CUDA and then manually copied CUDNN files?

**Reply by jjyyxx** [reply]
cudnn can also installed via apt, at least on my machines.

**Comment by hawkinsp** [comment]
You need to ensure that your CUDA libraries are in your `LD_LIBRARY_PATH`. Try setting the `LD_LIBRARY_PATH` environment variable.

**Comment by tgautam03** [comment, marked as answer]
I was able to install CUDA 11.7 using the following commands:
```
wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/cuda-ubuntu2004.pin
sudo mv cuda-ubuntu2004.pin /etc/apt/preferences.d/cuda-repository-pin-600
sudo apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub
sudo add-apt-repository "deb https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/ /"
sudo apt-get update
sudo apt-get -y install cuda
```

I then copied CUDNN files in `usr/local/cuda-11.7` and everything works ok. I wish all this was there in JAX installation guide for Ubuntu (for beginners like me).

**Reply by hawkinsp** [reply]
We actually have a better plan coming soon, which is we're hoping to use `pip` to install CUDA and CuDNN. So a simple `pip install` will install everything. Not yet, though!

**Reply by jjyyxx** [reply]
A system-wide cuda installation avoids bloated wheel size (like PyTorch) though.

**Reply by hawkinsp** [reply]
No, I mean: JAX will depend on the separate NVIDIA-provided CUDA wheels. The JAX wheel itself won't change size. (e.g., https://pypi.org/project/nvidia-cudnn-cu11/)
